build/tasks/AnalyzeBuildGraph.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using RepoTasks.ProjectModel;
+using RepoTasks.Utilities;
+
+namespace RepoTasks
+{
+    public class AnalyzeBuildGraph : Task, ICancelableTask
+    {
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+
+        /// <summary>
+        /// Repositories that we are building new versions of.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Solutions { get; set; }
+
+        [Required]
+        public string Properties { get; set; }
+
+        /// <summary>
+        /// New packages we are compiling. Used in the pin tool.
+        /// </summary>
+        [Output]
+        public ITaskItem[] PackagesProduced { get; set; }
+
+        /// <summary>
+        /// The order in which to build repositories
+        /// </summary>
+        [Output]
+        public ITaskItem[] RepositoryBuildOrder { get; set; }
+
+        public void Cancel()
+        {
+            _cts.Cancel();
+        }
+
+        public override bool Execute()
+        {
+            var factory = new SolutionInfoFactory(Log, BuildEngine5);
```

Comment:
Out of curiosity, what is BuildEngine5?

Reply:
Provides access to MSBuild's internal memory caching mechanisms.